pyevmasm/evmasm.py

```diff
@@ -390,13 +411,54 @@ def assemble_all(asmcode, pc=0, fork=DEFAULT_FORK):
     """
     asmcode = asmcode.split('\n')
     asmcode = iter(asmcode)
+ 
+    # we use a dictionary to record label locations:
+    labels = {}
+    # another dictionary to record which instruction
+    # we need to fill in.
+    fillins = {}
+    # we have to traverse the generated instruction twice
+    # so no use of generator here
+    instrs = []
+
     for line in asmcode:
-        if not line.strip():
+        line = line.strip()
+
+        # skip empty lines
+        if not line:
             continue
-        instr = assemble_one(line, pc=pc, fork=fork)
-        yield instr
+
+        # remove comments
+        index = line.find("#")
+        if index is not -1:
+            line = line[:index]
+
+        # skip directives:
+        if line.find(".") is 0:
+            continue
+
+        # handle labels
+        if line.endswith(":"):
+            # this is a label, record it with location (PC)
+            labels[line[:-1]] = pc
+            continue
+
+        instr = assemble_one(line, pc=pc, fork=fork, fillins=fillins)
+        instrs.append(instr)
         pc += instr.size
 
+    # fixup instructions
+    for label in labels:
+        if label not in fillins.keys():
+            continue
+        for instr in instrs:
```

Comment:
Instead of iterating over al linstructions for each label that qualifies maybe just rework all this do a single iteration over the instructions and fix the ones that needs it ? The instructions with `Instruction.uses_labels()` true